contracts/ronin/validator/CoinbaseExecution.sol

```diff
@@ -30,28 +30,41 @@ abstract contract CoinbaseExecution is
   using EnumFlags for EnumFlags.ValidatorFlag;
 
   modifier onlyCoinbase() {
-    if (msg.sender != block.coinbase) revert ErrCallerMustBeCoinbase();
+    _requiresCoinbase();
     _;
   }
 
+  function _requiresCoinbase() internal view {
+    assembly {
+      if iszero(eq(caller(), coinbase())) {
+        /// @dev value is equal to bytes4(keccak256("ErrCallerMustBeCoinbase()"))
+        mstore(0x00, 0x09f358fd)
+        revert(0x1c, 0x04)
+      }
+    }
+  }
+
   modifier whenEpochEnding() {
     if (!epochEndingAt(block.number)) revert ErrAtEndOfEpochOnly();
     _;
   }
 
   modifier oncePerEpoch() {
+    _oncePerEpoch();
+    _;
+  }
+
+  function _oncePerEpoch() internal {
     if (epochOf(_lastUpdatedBlock) >= epochOf(block.number)) revert ErrAlreadyWrappedEpoch();
     _lastUpdatedBlock = block.number;
-    _;
   }
 
   /**
    * @inheritdoc ICoinbaseExecution
    */
   function submitBlockReward() external payable override onlyCoinbase {
     bool _requestForBlockProducer = isBlockProducer(msg.sender) &&
-      !_jailed(msg.sender) &&
-      !_miningRewardDeprecated(msg.sender, currentPeriod());
+      !(_jailed(msg.sender) || _miningRewardDeprecated(msg.sender, currentPeriod()));
```

Comment:
Need double check on this conversion. IMO, coding is mainly for readability. The new code seems to be harder to understand.

Reply:
!a && !b cost 3 opcode: 2 "!" and 1 "&&"
!(a || b) cost 2 opcode: 1 "!" and 1 "||"

above modification save: 0.003kb